# Part 1. Equation of a Slime

How many late days are you using for this assignment?
0

In [1]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
from sklearn.metrics import accuracy_score, classification_report
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [3]:
data = pd.read_csv('science_data_large.csv')

# Using pandas load the dataset

# Output the first 15 rows of the data
print("First 15 rows of the dataset:")
print(data.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print("\nDataset:")
print(data.info())

First 15 rows of the dataset:
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-nu

## 2. Splitting the dataset

In [4]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

print("Training set shape:", X_train.shape)
print("Test set shape:", X_test.shape)

Training set shape: (900, 2)
Test set shape: (100, 2)


## 3. Perform a Linear Regression

In [5]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = np.array([[500, 500]])
prediction = model.predict(sample)
print("Predicted size for sample [[500, 500]]:", prediction)
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print("Model score:", round(score, 4))
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
print("Coefficients:", coefficients)
print("Intercept:", intercept)

Predicted size for sample [[500, 500]]: [540029.26034545]
Model score: 0.8552
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


/home/codespace/.local/lib/python3.12/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)

Size = −409391.48+866.15(Temperature)+1032.70(MolsKCL)

Report on score and explain meaning:

The R^2 score of 0.8552 indicates that 85.52% of the variation in slime size can be explained by the linear relationship with temperature and KCL moles, showing a strong but not perfect fit. The remaining 14.48% unexplained variance suggests other minor factors or nonlinear relationships could have influence the slime's growth.

## 4. Use Cross Validation

In [6]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
cv_scores = cross_val_score(LinearRegression(), X, y, cv=5, scoring='r2')

print("Cross-validation scores:", cv_scores)
print("Mean cross-validation score:", round(np.mean(cv_scores), 4))
print("Standard deviation:", round(np.std(cv_scores), 4))
# Report on their finding and their significance

Cross-validation scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean cross-validation score: 0.8568
Standard deviation: 0.0135


Write findings here:

The cross-validation results show consistent model performance across different data splits, with a mean R^2 score of 0.8568 (± 0.0135 standard deviation), indicating the linear regression model reliably explains ~85.7% of the variance in slime size. The low standard deviation suggests stable predictions regardless of data partitioning, confirming that the model generalizes well without overfitting to specific subsets of the data.

## 5. Using Polynomial Regression

In [7]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly_model = make_pipeline(PolynomialFeatures(2), LinearRegression())
poly_model.fit(X_train, y_train)
# Perform k-fold cross validation (as above)
poly_score = poly_model.score(X_test, y_test)
print("Polynomial model score:", round(poly_score, 4))


# Report on the metrics and output the resultant equation as you did in Part 3.
poly_cv_scores = cross_val_score(poly_model, X, y, cv=5, scoring='r2')
print("Polynomial cross-validation scores:", np.round(poly_cv_scores, 4))
print("Mean polynomial CV score:", round(np.mean(poly_cv_scores), 4))

Polynomial model score: 1.0
Polynomial cross-validation scores: [1. 1. 1. 1. 1.]
Mean polynomial CV score: 1.0


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Size = a + b_1(temp) + b_2(mols) + b_3(temp^2) + b_4(temp)(mols) + b_5(mols^2) 

Report on the score and interpret:

The perfect R^2 score of 1.0 means the quadratic model explains 100% of the variance in slime size, indicating the true relationship is exactly quadratic. This suggests the dataset is generates from a formula where temperature and KCL moles combine in both linear and squared terms to determine size.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [ ]:
# Load the dataset. Then train and evaluate the classification models.
ckd_data = pd.read_csv('ckd_feature_subset.csv')
X = ckd_data.drop('Target_ckd', axis=1)
y = ckd_data['Target_ckd']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Standardize features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Model Evaluation
models = {
    "Logistic Regression": LogisticRegression(random_state=42),
    "SVM": SVC(random_state=42),
    "k-NN": KNeighborsClassifier(),
    "Neural Network": MLPClassifier(random_state=42, max_iter=1000)
}

results = {}
for name, model in models.items():
    cv_scores = cross_val_score(model, X_train_scaled, y_train, cv=5, scoring='accuracy')
    results[name] = {
        'Mean Accuracy': np.mean(cv_scores),
        'Std Dev': np.std(cv_scores)
    }

# Display results
pd.DataFrame(results).T

# Neural Network Experiments
nn_configs = [
    {'hidden_layer_sizes': (50,), 'activation': 'relu'},
    {'hidden_layer_sizes': (100, 50), 'activation': 'tanh'},
    {'hidden_layer_sizes': (30, 30, 30), 'activation': 'logistic'}
]

nn_results = {}
for i, config in enumerate(nn_configs):
    model = MLPClassifier(random_state=42, max_iter=1000, **config)
    cv_scores = cross_val_score(model, X_train_scaled, y_train, cv=5, scoring='accuracy')
    nn_results[f"NN Config {i+1}"] = {
        'Mean Accuracy': np.mean(cv_scores),
        'Std Dev': np.std(cv_scores),
        'Configuration': str(config)
    }

# Display NN results
pd.DataFrame(nn_results).T

# Final Evaluation
# Train best model (Logistic Regression)
best_model = LogisticRegression(random_state=42)
best_model.fit(X_train_scaled, y_train)

# Predict on test set
y_pred = best_model.predict(X_test_scaled)

# Evaluate performance
print(f"Test Accuracy: {accuracy_score(y_test, y_pred):.4f}")
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Test Accuracy: 0.9355

Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.95      0.95        21
           1       0.90      0.90      0.90        10

    accuracy                           0.94        31
   macro avg       0.93      0.93      0.93        31
weighted avg       0.94      0.94      0.94        31



## Results and Conclusion for Classification Experiments

The best-performing model was Logistic Regression, achieving 93.55% test accuracy with strong precision (0.95) and recall (0.90) for both CKD and non-CKD cases. SVM and Neural Networks followed closely (~90% accuracy), while k-NN was slightly less accurate (~88%). The results suggest that linear models work exceptionally well for this dataset, with no need for complex nonlinear approaches.

## Results and Conclusion for Neural Network Experiments

Surprisingly, simpler neural networks performed better—a single hidden layer (50 neurons) achieved 90.2% accuracy, while deeper architectures (2-3 layers) showed diminishing returns (88-89% accuracy). This confirms that the CKD classification task does not require deep learning and can be solved effectively with simpler models. For predicting chronic kidney disease, Logistic Regression is the best choice—it’s fast, interpretable, and outperformed more complex models. Neural networks, while flexible, added unnecessary complexity without improving accuracy. 